torch/lib/c10d/ProcessGroupNCCL.hpp

```diff
@@ -143,66 +144,129 @@ class ProcessGroupNCCL : public ProcessGroup {
+    // Just returns FutureNCCL's value.
     at::IValue value() override {
-      work_->wait();
-      return outputs_;
+      return value_;
     }
 
     const at::IValue& constValue() override {
-      work_->wait();
-      return outputs_;
+      return value_;
     }
 
-    // Add a callback to FutureNCCL. It invokes the callback inline after
-    // WorkNCCL's wait(). workCallbacks return a Future (not FutureNCCL).
-    void addCallback(std::function<void(void)> callback) override {
-      work_->wait();
+    // Adds a callback to FutureNCCL. It invokes the callback inline after
+    // synchronizing FutureNCCL's own cudaEvent with the stream that runs
+    // this callback. FutureNCCL then will return a FutureNCCL if this callback
+    // is executed without any errors. This new FutureNCCL's cudaEvent will
+    // record the callback's stream.
+    void addCallbackWithStream(
+        std::function<void(void)> callback,
+        const c10::cuda::CUDAStream stream,
+        std::shared_ptr<at::cuda::CUDAEvent> thenFutCudaEvent) {
+      (*cudaEvent_).block(stream);
+      c10::OptionalStreamGuard streamGuard{c10::Stream(stream)};
       callback();
+      (*thenFutCudaEvent).record(stream);
+    }
+
+    // We use addCallbackWithStream instead of addCallback.
+    void addCallback(std::function<void(void)> /* unused */) override {
+      C10_THROW_ERROR(
+          Error,
+          "FutureNCCL uses addCallbackWithStream instead of addCallback.");
+    }
+
+    // Adds a callback to the futureNCCL, and returns another FutureNCCL to hold
+    // the return value of the callback and new cudaEvent that recorded the
+    // stream that runs this callback.
+    c10::intrusive_ptr<Future> then(
+        std::function<at::IValue(void)> callback,
+        at::TypePtr type) override {
+      // Set a default fut, this will be returned in case of an error in the
+      // callback. If the callback is executed without any errors, fut will be
+      // replaced by a FutureNCCL.
+      auto fut = c10::make_intrusive<Future>(type);
+
+      // Get a new stream from pool that will run the callback.
+      const c10::cuda::CUDAStream stream =
+          at::cuda::getStreamFromPool(deviceIndex_);
+      // Create a new cudaEvent object that will record callback's stream and
+      // used by the new FutureNCCL.
+      auto thenFutCudaEvent = std::make_shared<at::cuda::CUDAEvent>();
+
+      // Cannot move capture std::function in lambda, because it cannot deduce
+      // the template type for std::function. Hence use std::bind to explicitly
+      // specify types.
+      addCallbackWithStream(
+          std::bind(
+              [&](std::function<at::IValue(void)> cb) {
+                try {
+                  fut = c10::make_intrusive<FutureNCCL>(
+                      cb(), deviceIndex_, thenFutCudaEvent);
+                } catch (const std::exception& e) {
+                  fut->setError(e.what());
```

Comment:
You probably need to override `setError` in FutureNCCL to ensure this is handled appropriately. Also, `wait()`, `value()` and `constValue()` need to look for an error being set.